File: leetcode/intervals/435. Non-overlapping Intervals/js/main.js
// Time O(NLogN)
// Space O(N)

/**
 * @param {number[][]} intervals
 * @return {number}
 */
const eraseOverlapIntervals = intervals => {
  if (intervals.length == 0) return 0;

  intervals.sort((a, b) => a[0] - b[0]);

  let [start, end] = intervals[0];
  let res = 0;

  for (let i = 1; i < intervals.length; i++) {
    if (overlap([start, end], intervals[i])) {
      start = Math.max(start, intervals[i][0]);
      end = Math.min(end, intervals[i][1]);
      res++;
    } else {
      start = intervals[i][0];
      end = intervals[i][1];
    }
  }

  return res;

  function overlap([x, y], [u, z]) {
    return z > x && y > u;
  }
};

File: leetcode/intervals/435. Non-overlapping Intervals/py/main.py
# Time O(NLogN)
# Space O(N)
class Solution:
    def eraseOverlapIntervals(self, intervals):
        if len(intervals) == 0:
          return 0

        intervals.sort(key = lambda x: (x[0], -x[1]))

        start, end = intervals[0]
        res = 0

        for i in range(1, len(intervals)):
          if self.overlap([start, end], intervals[i]):
            start = max(start, intervals[i][0])
            end = min(end, intervals[i][1])
            res += 1
          else:
            start, end = intervals[i]
        return res
        
    def overlap(self, int1, int2):
       [x, y] = int1
       [u, z] = int2
       return z > x and y > u

